File: csharp/src/detectors/untrusted-deserialization/UntrustedDeserialization.cs
using Newtonsoft.Json;

namespace InsecureDeserialization
{
    class UntrustedDeserialization
    {
        // {fact rule=untrusted-deserialization@v1.0 defects=1}
        public void UntrustedDeserializationNoncompliant(String jsonpath)
        {
            JsonConvert.DeserializeObject<object>(jsonpath, new JsonSerializerSettings
                {
                    // Noncompliant: `TypeNameHandling.All` is not safe.
                    TypeNameHandling = TypeNameHandling.All
                }
            );
        }
        // {/fact}
        
        // {fact rule=untrusted-deserialization@v1.0 defects=0}
        public void UntrustedDeserializationCompliant(String jsonpath)
        {
            JsonConvert.DeserializeObject<object>(jsonpath, new JsonSerializerSettings
            {
                // Compliant: `TypeNameHandling.None` is safe.
                TypeNameHandling = TypeNameHandling.None
                }
            );
        }
        // {/fact}
    }
}

File: ruby/src/detectors/untrusted-deserialization/untrusted-deserialization.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=untrusted-deserialization@v1.0 defects=1}
def handler_noncompliant(event:, context:)
  foobar = event['smth']
  # Noncompliant: tainted data passed inside `CSV.load()`.
  obj3 = CSV.load("o:" + event['data'])
end
# {/fact}

# {fact rule=untrusted-deserialization@v1.0 defects=0}
def handler_compliant(event:, context:)
  # Compliant: no tainted data passed inside `CSV.load()`.
  obj3 = CSV.load(get_safe_data())
end
# {/fact}